### ISSUE TITLE
PyMySQL 0.7.7 is injecting strange codes into exception messages

### ISSUE THREAD

**Issue by zzzeek** [PyMySQL/PyMySQL#507]
Exception messages have changed as of PyMySQL 0.7.7 such that some kind of codes are present right at the start of the message.   This breaks code that is looking for text within the message to distinguish the exact kind of error this is as well as parsing out the arguments.

Example, the following two errors in 0.7.6 look like this:

     (1054, u"Unknown column 'fake' in 'field list'")
    (1452, u'Cannot add or update a child row: a foreign key constraint fails (`test`.`b`, CONSTRAINT `b_ibfk_1` FOREIGN KEY (`a_id`) REFERENCES `a` (`id`))')

in 0.7.7, we get:

    (1054, u"42S22Unknown column 'fake' in 'field list'")
    (1452, u'23000Cannot add or update a child row: a foreign key constraint fails (`test`.`b`, CONSTRAINT `b_ibfk_1` FOREIGN KEY (`a_id`) REFERENCES `a` (`id`))')

those codes seem to be fixed so perhaps they are some kind of debugging string that forgot to be removed, but this is breaking a lot of downstream stuff.

test case:

```
import pymysql
import warnings

warnings.simplefilter("ignore")


conn = pymysql.connect(user="scott", passwd="tiger", host="127.0.0.1", db="test")

cursor = conn.cursor()


def show_exception(fn, *args):
    try:
        fn(*args)
    except Exception as err:
        print(err)

show_exception(
    cursor.execute, "select fake"
)


cursor.execute("""
    CREATE TABLE IF NOT EXISTS a (
       id INTEGER PRIMARY KEY
    ) ENGINE=InnoDB
""")

cursor.execute("""
    CREATE TABLE IF NOT EXISTS b (
       id INTEGER PRIMARY KEY,
       a_id INTEGER,
       FOREIGN KEY (a_id) REFERENCES a(id)
    ) ENGINE=InnoDB
""")

show_exception(
    cursor.execute, "INSERT INTO b (id, a_id) VALUES (1, 25)"
)
```

**Comment by dianaclarke** [comment]
It's the removal of the following `is_41` code block in commit https://github.com/PyMySQL/PyMySQL/commit/02910b74bb4f776b0a71d0ab71d9a5b99273685e that's behind the exception message changes.

https://github.com/PyMySQL/PyMySQL/blob/c3ed30135117a13d5e52087d67249a252317c30a/pymysql/err.py#L100

Those additional numbers are the `sqlstate`.

**Comment by dianaclarke** [comment]
Here's a pull request for issue #508. Do let me know if you want it changed in any way. Thanks!

**Comment by methane** [comment]
I'm sorry about it.
I didn't expect people use "error message for human" even though there is error code.
Why do you use error message for human instead of error code?

I use Exception type (IOError) or value (e.errno) and don't expect "error message for human"
is stable enough for program logic.
Error message for human may be changed to more user friendly, descriptive, typo-fixed
string in any version up.
In this case, I felt stripping sqlstate is meaningless.

I'll release 0.7.8 tomorrow if possible.

**Comment by dianaclarke** [comment]
No need to apologize. Thanks for all you do!

PS. And thanks for the reviews!

**Comment by zzzeek** [comment]
@methane because  1. there's a lot more detail embedded in the message that we need 2. this is part of a system that works across DBAPIs and they don't all have error codes, see https://github.com/openstack/oslo.db/blob/master/oslo_db/sqlalchemy/exc_filters.py#L191.    The Python DBAPI unfortunately has no specification for specific error situations, which kinds of exceptions should be used in what case (example, lots of OperationalErrors in mysqlclient are InternalErrors in pymysql) or what should be captured in the object.

We've worked around this issue downstream already by changing our regexp.

**Comment by zzzeek** [comment]
@methane I will grant that the whole "we need to capture the constraint name, table name, column name" etc. is a requirement that preceded my working on Openstack.   I can try to start encouraging that this feature be phased out because i doubt it's needed by anyone.

**Comment by jd** [comment]
@zzzeek actually it's very useful to convert integrity error to proper Python exception (see https://github.com/openstack/gnocchi/blob/master/gnocchi/indexer/sqlalchemy.py#L584 for example)

**Comment by methane** [comment]
@zzzeek @jd I see, thanks.
I won't change it in 0.8 and 1.0.

**Comment by methane** [comment]
But I'll change InternalError to OperationalError when I realize it.  Are you OK about it?

**Comment by methane** [comment]
released 0.7.8

**Comment by zzzeek** [comment]
@methane that's probably OK I think the InternalError situation is in migration scripts where they try to create an INDEX that already exists or something.  I misspoke when I said "lots".
